tools/c7n_azure/c7n_azure/policy.py

```diff
@@ -0,0 +1,80 @@
+# Copyright 2015-2018 Capital One Services, LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from c7n_azure.template_utils import TemplateUtilities
+
+from c7n import utils
+from c7n.policy import ServerlessExecutionMode, execution
+
+
+@execution.register('azure-function')
+class AzureFunctionMode(ServerlessExecutionMode):
+    """A policy that runs/executes in azure functions."""
+
+    azure_function_schema = {
+        'type': 'object',
+        'additionalProperties': False,
+        'properties': {
+            'provision-options': {
+                'type': 'object',
+                'sku': 'string',
+                'location': 'string',
+                'appInsightsLocation': 'string',
+                'workerSize': 'number',
+                'skuCode': 'string'
+            }
+        }
+    }
+
+    schema = utils.type_schema('azure-function', rinherit=azure_function_schema)
+
+    POLICY_METRICS = ('ResourceCount', 'ResourceTime', 'ActionTime')
+
+    def __init__(self, policy):
+        self.policy = policy
+        self.template_util = TemplateUtilities()
+
+    def run(self, event=None, lambda_context=None):
+        """Run the actual policy."""
+        raise NotImplementedError("subclass responsibility")
+
+    def provision(self):
+        """Provision any resources needed for the policy."""
+        parameters = self.get_parameters()
+        group_name = parameters['name']['value']
+
+        self.template_util.create_resource_group(
+            group_name, {'location': parameters['location']['value']})
+
+        self.template_util.deploy_resource_template(
```

Comment:
We have to wait on it to finish before we can do the next step of pushing the action function, so we sort of have to be sync in this case.  The way I was thining about it this would rarely be used - we only need one app service in the subscription which all functions can go in (and the user can make it themselves in advance if they want full control) - so the general case should be the policy just listing the app service name to deploy to I think, or at least us ignoring the provisioning step.